From: Tias Guns <tias@ulyssis.org>
Date: Sun, 5 Aug 2012 16:13:55 +0200
Subject: [PATCH] ash history

allows ash history to work on Android

Patch modified from 'busybox-android.patch' by Alexandre Courbot
https://github.com/Gnurou/busybox-android
---
 include/libbb.h |    4 ++--
 init/init.c     |    2 +-
 shell/ash.c     |    8 +++++++-
 3 files changed, 10 insertions(+), 4 deletions(-)

Index: src/include/libbb.h
===================================================================
--- src.orig/include/libbb.h
+++ src/include/libbb.h
@@ -2038,12 +2038,12 @@ extern struct globals *const ptr_to_glob
  * use bb_default_login_shell and following defines.
  * If you change LIBBB_DEFAULT_LOGIN_SHELL,
  * don't forget to change increment constant. */
-#define LIBBB_DEFAULT_LOGIN_SHELL  "-/bin/sh"
+#define LIBBB_DEFAULT_LOGIN_SHELL  "-/sbin/sh"
 extern const char bb_default_login_shell[] ALIGN1;
 /* "/bin/sh" */
 #define DEFAULT_SHELL              (bb_default_login_shell+1)
 /* "sh" */
-#define DEFAULT_SHELL_SHORT_NAME   (bb_default_login_shell+6)
+#define DEFAULT_SHELL_SHORT_NAME   (bb_default_login_shell+7)
 
 /* The following devices are the same on all systems.  */
 #define CURRENT_TTY "/dev/tty"
Index: src/init/init.c
===================================================================
--- src.orig/init/init.c
+++ src/init/init.c
@@ -1101,7 +1101,7 @@ int init_main(int argc UNUSED_PARAM, cha
 	/* Make sure environs is set to something sane */
 	putenv((char *) "HOME=/");
 	putenv((char *) bb_PATH_root_path);
-	putenv((char *) "SHELL=/bin/sh");
+	putenv((char *) "SHELL=/sbin/sh");
 	putenv((char *) "USER=root"); /* needed? why? */
 
 	if (argv[1])
Index: src/shell/ash.c
===================================================================
--- src.orig/shell/ash.c
+++ src/shell/ash.c
@@ -13921,7 +13921,11 @@ int ash_main(int argc UNUSED_PARAM, char
 		const char *hp;
 
 		state = 1;
+#if !defined(__ANDROID__) && !defined(ANDROID)
 		read_profile("/etc/profile");
+#else
+		read_profile("/system/etc/profile");
+#endif
  state1:
 		state = 2;
 		hp = lookupvar("HOME");
@@ -13959,6 +13963,10 @@ int ash_main(int argc UNUSED_PARAM, char
 		if (iflag) {
 			const char *hp = lookupvar("HISTFILE");
 			if (!hp) {
+#ifdef __ANDROID__
+				setvar("HISTFILE", "/mnt/sdcard/ash_history", 0);
+				hp = lookvar("HISTFILE");
+#else
 				hp = lookupvar("HOME");
 				if (hp) {
 					INT_OFF;
@@ -13968,6 +13976,7 @@ int ash_main(int argc UNUSED_PARAM, char
 					INT_ON;
 					hp = lookupvar("HISTFILE");
 				}
+#endif
 			}
 			if (hp)
 				line_input_state->hist_file = hp;
